File: backend/src/main/java/com/example/user/application/service/UpdateUserService.java
package com.example.user.application.service;

import com.example.user.application.command.UpdateUserCommand;
import com.example.user.application.command.UserResult;
import com.example.user.application.mapper.UserApplicationMapper;
import com.example.user.application.port.in.UpdateUserUseCase;
import com.example.user.domain.exception.UserNotFoundException;
import com.example.user.domain.model.User;
import com.example.user.domain.port.UserRepository;

/**
 * Updates user profile and role.
 */
public final class UpdateUserService implements UpdateUserUseCase {

    private final UserRepository userRepository;
    private final UserApplicationMapper userApplicationMapper;

    public UpdateUserService(UserRepository userRepository, UserApplicationMapper userApplicationMapper) {
        this.userRepository = userRepository;
        this.userApplicationMapper = userApplicationMapper;
    }

    @Override
    public UserResult execute(UpdateUserCommand command) {
        User user = userRepository
                .findById(command.getId())
                .orElseThrow(() -> new UserNotFoundException(command.getId()));
        user.update(command.getFirstName(), command.getLastName(), command.getRole(), command.getStatus());
        User saved = userRepository.save(user);
        return userApplicationMapper.toResult(saved);
    }
}

File: frontend/src/user/application/service/UpdateUserService.ts
import { BaseMutationService } from '@/shared/application/service/BaseMutationService';
import { UserAdminGateway, UserDto } from '@/user/domain/port/UserAdminGateway';
import { UpdateUserUseCase } from '@/user/application/port/in/UpdateUserUseCase';
import { UpdateUserCommand } from '@/user/application/command/UpdateUserCommand';

export class UpdateUserService extends BaseMutationService<UpdateUserCommand, UserDto> implements UpdateUserUseCase {
  constructor(private readonly gateway: UserAdminGateway) {
    super();
  }

  protected handleMutation(command: UpdateUserCommand): Promise<UserDto> {
    return this.gateway.update(command.id, command.payload);
  }
}
